service/storage_proxy.cc

```diff
@@ -5960,6 +5959,97 @@ storage_proxy::query_partition_key_range(lw_shared_ptr<query::read_command> cmd,
+    auto p = shared_from_this();
+    auto& cf = _db.local().find_column_family(schema);
+    auto pcf = _db.local().get_config().cache_hit_rate_read_balancing() ? &cf : nullptr;
+    const auto& tm = erm->get_token_metadata();
+
+    if (_features.range_scan_data_variant) {
+        cmd->slice.options.set<query::partition_slice::option::range_scan_data_variant>();
+    }
+
+    auto& preferred_replicas = query_options.preferred_replicas;
+
+    const auto preferred_replicas_for_range = [&preferred_replicas, &tm] (const dht::partition_range& r) {
+        auto it = preferred_replicas.find(r.transform(std::mem_fn(&dht::ring_position::token)));
+        return it == preferred_replicas.end() ? inet_address_vector_replica_set{} : replica_ids_to_endpoints(tm, it->second);
+    };
+    const auto to_token_range = [] (const dht::partition_range& r) { return r.transform(std::mem_fn(&dht::ring_position::token)); };
+
+    dht::partition_range_vector ranges = ranges_to_vnodes(1);
+    dht::partition_range_vector::iterator i = ranges.begin();
+
+    if (i != ranges.end()) {
+        dht::partition_range& range = *i;
+        inet_address_vector_replica_set live_endpoints = get_endpoints_for_reading(schema->ks_name(), *erm, end_token(range));
+        inet_address_vector_replica_set preferred_replicas = preferred_replicas_for_range(*i);
+        inet_address_vector_replica_set filtered_endpoints = filter_replicas_for_read(cl, *erm, live_endpoints, preferred_replicas, pcf);
+        auto token_range = to_token_range(range);
+
+        slogger.trace("creating range read executor for range {} in table {}.{} with targets {}",
+                    range, schema->ks_name(), schema->cf_name(), filtered_endpoints);
+        try {
+            db::assure_sufficient_live_nodes(cl, *erm, filtered_endpoints);
+        } catch (exceptions::unavailable_exception& ex) {
+            slogger.debug("Read unavailable: cl={} required {} alive {}", ex.consistency, ex.required, ex.alive);
+            get_stats().range_slice_unavailables.mark();
+            throw;
+        }
+
+        auto exec = ::make_shared<never_speculating_read_executor>(schema, cf.shared_from_this(), p, erm, cmd, std::move(range), cl, std::move(filtered_endpoints), trace_state, permit, std::monostate());
+        auto wrapped_result = co_await exec->execute(timeout);
+
+        if (!wrapped_result) {
+            auto error = std::move(wrapped_result).assume_error();
+            handle_read_error(error.clone(), true);
+            co_return error;
+        }
+
+        foreign_ptr<lw_shared_ptr<query::result>> result = std::move(wrapped_result).value();
+        result->ensure_counts();
+
+        auto used_replicas = replicas_per_token_range();
+        auto replica_ids = endpoints_to_replica_ids(tm, exec->used_targets());
+        used_replicas.emplace(token_range, replica_ids);
+
+        co_return coordinator_query_result(std::move(result), std::move(used_replicas));
```

Comment:
You should only return here if read was short or if this was the last range. Otherwise you should read the next tablet. I think this `if` should change to a loop, which stops when the current read was either short, or the page-limit reached 0 (you decrement on every read, by the amount of data read).

Reply:
The goal is to simplify here. The client already has a loop around pages.

Reply:
The pager's stop condition is a page which is not full and not short. To instruct the pager to continue, you have to set the short_read flag on the returned result. As far as the pager is concerned, the `short_read` flag doesn't have any special meaning, beyond: "continue reading despite incomplete page".